Creating self signed certificate using openssl cli requires digest to be explicitly set

**Describe the bug**
Tying to generate a self signed certificate using the openssl CLI without digest parameter fails:

```bash
openssl req -x509 -key "$KEY" -outform PEM -subj "/C=DE/CN=test.internal"
```

When explicitly specifying the digest however it woks:

```bash
openssl req -x509 -key "$KEY" -outform PEM -sha256 -subj "/C=DE/CN=test.internal"
# OUTPUT
# -----BEGIN CERTIFICATE-----
# MIIEKzCCApOgAwIBAgIUXwKZcOzkcCySxPL6LOQ5T/zam6kwDQYJKoZIhvcNAQEL
# ...
# -----END CERTIFICATE-----
```

The reason seems to be that `p11prov_mech_by_mechanism` is called with `CK_MECHANISM_TYPE mechanism = 0` because `sigctx->digest` was not set.


**To Reproduce**
Clone commit d58bc51dc96658290e5860be07957764588f7dcc (current master), build and run commands above

**Expected behavior**
Certificate should be generated, just like when pkcs11-provider is not being used. Otherwise documenting this behavior would be nice.

```bash
openssl req -x509 -new -nodes -outform PEM -subj "/C=DE/CN=test.internal"
# Output
# ...
# -----
# -----BEGIN CERTIFICATE-----
# MIIDKzCCAhOgAwIBAgIUY2Y/3FvdJ5woFzxzPGXSRD0qQb4wDQYJKoZIhvcNAQEL
# ...
# WRd67YZhEbHw6GCxlYEtSV0FmozxhP1vN+MdbGpxWb2aIAIyyVUxCpmlKyGLSu8=
# -----END CERTIFICATE-----
```




**Operating environment (please complete the following information):**
 - OS: Ubuntu
 - Version: Lunar / 23.04

**Token and application used (please complete the following information):**
 - Device: SoftHSMv2
 - PKCS11 Driver version: 2.6.1
 - Application: openssl CLI
 - Version: `OpenSSL 3.0.8 7 Feb 2023 (Library: OpenSSL 3.0.8 7 Feb 2023)`

**Additional context**

The openssl CLI docs says something along the lines of

If not specified then SHA1 is used with -fingerprint or the default digest for the signing algorithm is used, typically SHA256.

Ah I know why I could not repdocue I have this in openssl.cnf:
```
[ req ]
default_bits            = 2048
default_md             = sha256
```

If I comment out the default I get:
```
 openssl req -x509 -key $ECPRIURI -outform PEM -subj "/C=DE/CN=test.internal"
00F22CEE3C7F0000:error:40800012:pkcs11:p11prov_GetAttributeValue:Invalid attribute type specified in a template:interface.gen.c:408:Error returned by C_GetAttributeValue
00F22CEE3C7F0000:error:40800012:pkcs11:p11prov_GetAttributeValue:Invalid attribute type specified in a template:interface.gen.c:408:Error returned by C_GetAttributeValue
00F22CEE3C7F0000:error:40800012:pkcs11:p11prov_GetAttributeValue:Invalid attribute type specified in a template:interface.gen.c:408:Error returned by C_GetAttributeValue
00F22CEE3C7F0000:error:40800012:pkcs11:p11prov_GetAttributeValue:Invalid attribute type specified in a template:interface.gen.c:408:Error returned by C_GetAttributeValue
```

Those errors should also not be emitted as they have nothing to do with the issue at hand, but the point is that I do not et a cert either.

This is because we do not have a default digest set when initializing signatures.
I guess we could have one when digest-sign is used...

--- a/src/signature.c
+++ b/src/signature.c
@@ -1297,6 +1297,9 @@ static int p11prov_rsasig_get_ctx_params(void *ctx, OSSL_PARAM *params)
         case CKM_RSA_PKCS:
             result = p11prov_mech_by_mechanism(sigctx->digest, &mech);
             if (result != CKR_OK) {
+                P11PROV_raise(
+                    sigctx->provctx, result,
+                    "Failed to get digest for signature algorithm ID");
                 return RET_OSSL_ERR;
             }
             ret = OSSL_PARAM_set_octet_string(p, mech->der_rsa_algorithm_id,
@@ -1353,12 +1356,13 @@ static int p11prov_rsasig_get_ctx_params(void *ctx, OSSL_PARAM *params)
     p = OSSL_PARAM_locate(params, OSSL_SIGNATURE_PARAM_MGF1_DIGEST);
     if (p) {
         const char *digest = NULL;
+        CK_RV rv = CKR_GENERAL_ERROR;
 
         if (sigctx->pss_params.mgf != 0) {
             digest = p11prov_sig_mgf_name(sigctx->pss_params.mgf);
         } else {
             const P11PROV_MECH *pssmech;
-            CK_RV rv = p11prov_mech_by_mechanism(sigctx->mechtype, &pssmech);
+            rv = p11prov_mech_by_mechanism(sigctx->mechtype, &pssmech);
             if (rv == CKR_OK) {
                 rv = p11prov_digest_get_name(pssmech->digest, &digest);
                 if (rv != CKR_OK) {
@@ -1367,6 +1371,7 @@ static int p11prov_rsasig_get_ctx_params(void *ctx, OSSL_PARAM *params)
             }
         }
         if (!digest) {
+            P11PROV_raise(sigctx->provctx, rv, "Failed to get digest for MGF1");
             return RET_OSSL_ERR;
         }
         ret = OSSL_PARAM_set_utf8_string(p, digest);
@@ -1929,6 +1934,9 @@ static int p11prov_ecdsa_get_ctx_params(void *ctx, OSSL_PARAM *params)
         case CKM_ECDSA:
             result = p11prov_mech_by_mechanism(sigctx->digest, &mech);
             if (result != CKR_OK) {
+                P11PROV_raise(
+                    sigctx->provctx, result,
+                    "Failed to get digest for signature algorithm ID");
                 return RET_OSSL_ERR;
             }
             ret = OSSL_PARAM_set_octet_string(p, mech->der_ecdsa_algorithm_id,

--- a/src/signature.c
+++ b/src/signature.c
@@ -1170,6 +1170,11 @@ static int p11prov_rsasig_digest_sign_init(void *ctx, const char *digest,
     P11PROV_debug("rsa digest sign init (ctx=%p, digest=%s, key=%p, params=%p)",
                   ctx, digest ? digest : "<NULL>", provkey, params);
 
+    /* use a default of sha2 256 if not provided */
+    if (!digest) {
+        digest = "sha256";
+    }
+
     ret = p11prov_sig_op_init(ctx, provkey, CKF_SIGN, digest);
     if (ret != CKR_OK) {
         return RET_OSSL_ERR;
@@ -1219,6 +1224,11 @@ static int p11prov_rsasig_digest_verify_init(void *ctx, const char *digest,
 
     P11PROV_debug("rsa digest verify init (ctx=%p, key=%p, params=%p)", ctx,
                   provkey, params);
+
+    /* use a default of sha2 256 if not provided */
+    if (!digest) {
+        digest = "sha256";
+    }
 
     ret = p11prov_sig_op_init(ctx, provkey, CKF_VERIFY, digest);
     if (ret != CKR_OK) {
@@ -1798,6 +1808,11 @@ static int p11prov_ecdsa_digest_sign_init(void *ctx, const char *digest,
         "ecdsa digest sign init (ctx=%p, digest=%s, key=%p, params=%p)", ctx,
         digest ? digest : "<NULL>", provkey, params);
 
+    /* use a default of sha2 256 if not provided */
+    if (!digest) {
+        digest = "sha256";
+    }
+
     ret = p11prov_sig_op_init(ctx, provkey, CKF_SIGN, digest);
     if (ret != CKR_OK) {
         return RET_OSSL_ERR;
@@ -1860,6 +1875,11 @@ static int p11prov_ecdsa_digest_verify_init(void *ctx, const char *digest,
 
     P11PROV_debug("ecdsa digest verify init (ctx=%p, key=%p, params=%p)", ctx,
                   provkey, params);
+
+    /* use a default of sha2 256 if not provided */
+    if (!digest) {
+        digest = "sha256";
+    }
 
     ret = p11prov_sig_op_init(ctx, provkey, CKF_VERIFY, digest);
     if (ret != CKR_OK) {

--- a/src/util.c
+++ b/src/util.c
@@ -23,6 +23,9 @@ CK_RV p11prov_fetch_attributes(P11PROV_CTX *ctx, P11PROV_SESSION *session,
         q[i] = attrs[i].attr;
     }
 
+    /* error stack mark so we can avoid returning bogus errors */
+    p11prov_set_error_mark(ctx);
+
     /* try one shot, then fallback to individual calls if that fails */
     ret = p11prov_GetAttributeValue(ctx, sess, object, q, attrnums);
     if (ret == CKR_OK) {
@@ -31,14 +34,18 @@ CK_RV p11prov_fetch_attributes(P11PROV_CTX *ctx, P11PROV_SESSION *session,
             if (q[i].ulValueLen == CK_UNAVAILABLE_INFORMATION) {
                 /* This can't happen according to the algorithm described
                  * in the spec when the call returns CKR_OK. */
-                return CKR_GENERAL_ERROR;
+                ret = CKR_GENERAL_ERROR;
+                P11PROV_raise(ctx, ret, "Failed to get attributes");
+                goto done;
             }
             if (attrs[i].allocate) {
                 /* always allocate one more, so that zero terminated strings
                  * work automatically */
                 q[i].pValue = OPENSSL_zalloc(q[i].ulValueLen + 1);
                 if (!q[i].pValue) {
-                    return CKR_HOST_MEMORY;
+                    ret = CKR_HOST_MEMORY;
+                    P11PROV_raise(ctx, ret, "Failed to get attributes");
+                    goto done;
                 }
                 /* add to re-request list */
                 r[retrnums] = q[i];
@@ -69,7 +76,9 @@ CK_RV p11prov_fetch_attributes(P11PROV_CTX *ctx, P11PROV_SESSION *session,
                     attrs[i].attr.pValue =
                         OPENSSL_zalloc(attrs[i].attr.ulValueLen + 1);
                     if (!attrs[i].attr.pValue) {
-                        return CKR_HOST_MEMORY;
+                        ret = CKR_HOST_MEMORY;
+                        P11PROV_raise(ctx, ret, "Failed to get attributes");
+                        goto done;
                     }
                 }
             }
@@ -91,6 +100,14 @@ CK_RV p11prov_fetch_attributes(P11PROV_CTX *ctx, P11PROV_SESSION *session,
                           attrs[i].attr.ulValueLen);
         }
         ret = CKR_OK;
+    }
+done:
+    if (ret == CKR_OK) {
+        /* if there was any error, remove it, as we got success */
+        p11prov_pop_error_to_mark(ctx);
+    } else {
+        /* otherwise clear the mark and leave errors on the stack */
+        p11prov_clear_last_error_mark(ctx);
     }
     return ret;
 }

